### PR TITLE
toxmm - toxcore with `sigc::signal` and `Glib::Property`

### PR DESCRIPTION
**Pull request by KoKuToru**
Full rewrite of the core of gTox.
Better now than later.

Because gTox is my first GTK+ App I ever wrote, I had no experiences..

So I started to rewrite the core part to use `sigc::signal` and `Glib::Property`, 
which simplifed the code and event handling a lot !

Right now I am updating everything to work with this `toxmm`.

Port-Status
=====
- [x] restructure folders
- [x] profile selection
- [x] profile create
  - [x] update, sqlite got removed
- [x] contact list
  - [x] contact list add contact
  - [x] contact list accept request
  - [x] contact list remove contact
- [x] chat
  - [ ] ability to close and destroy dialog/widget
  - [x] bubble
  - [x] message
  - [x] action
  - [x] style
 -  [x] file
    - [x] preview
    - [x] recv
    - [x] send
 -  [x] avatar
    - [x] recv
    - [x] send
- [x] remove sqlite
- [x] flatbuffers
  - [x] config
  - [x] log
  - [ ] use json-format
- [x] video

Do the following, when everything above is finish
---
- [x] port changes from master
- [x] remove old code
- [ ] make sure all cmake-scripts are running correctly
- [ ] update in code documentation of `toxmm::` and `utils::`
- [ ] update translation
- [ ] make sure everything is in the same coding style

### CODE REVIEW

File: src/dialog/error.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        add_button(_("ERROR_REPORT_BTN"), 213);
+    }
+}
+
+error::~error() {
+}
+
+int error::run() {
+    switch(MessageDialog::run()) {
+        case 213:
```
Comment (Spagy) [inline src/dialog/error.cpp]: What is 213?

File: src/dialog/error.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+error::error(Gtk::Window& parent,bool fatal,std::string title, std::string message):
+    MessageDialog(parent, title, false, fatal?Gtk::MESSAGE_ERROR:Gtk::MESSAGE_WARNING, Gtk::BUTTONS_CLOSE, true) {
+    set_secondary_text(((fatal)?_("ERROR_REPORT_TEXT_INTRO"):"")
+                       + Glib::Markup::escape_text(message)
+                       + ((fatal)?(Glib::Markup::escape_text("\n\n" + get_stacktrace())
+                                   + _("ERROR_REPORT_TEXT_OUTRO")):""), true);
+
+    if (fatal) {
+        add_button(_("ERROR_REPORT_BTN"), 213);
```
Comment (Spagy) [inline src/dialog/error.cpp]: Why 213 ?

Comment (KoKuToru) [inline src/dialog/error.cpp]: good question.. wtf ?

Comment (KoKuToru) [inline src/dialog/error.cpp]: Just checked documentation any positive number is okay,
It's used to check which button got pressed. at line 57.

I could replace it with a constexpr..

File: src/utils/dispatcher.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#ifndef DISPATCHER_H
+#define DISPATCHER_H
+#include <memory>
+#include <glibmm.h>
+
+namespace utils {
+    /**
+     * @brief The Dispatcher, executes given function on gtk main loop.
```
Comment (Spagy) [inline src/utils/dispatcher.h]: This looks like a doxygen tag but it's the only one I can see. Does gTox use doxygen? If not, what is this?

Comment (KoKuToru) [inline src/utils/dispatcher.h]: There is no doxygen, but there is more in the old `https://github.com/KoKuToru/gTox/blob/master/Source/Tox/Toxmm.h`...

I am using these doxygen-tags because my IDE displays these messages..
They are also visual different then normal comments.

I see normal comments as "not important" but nice to have.

I also tag todos with `//TODO: `, my current IDE doesn't highlight them, but others might..
(also `//HACK:` `//XXX:` `//FIXME:` ..)

But this rewrite is not finish yet, I will add the most important comments before merging it.

In my experince the generate doxygen will be hardly useful, we aren't writing a library..

but the whole `toxmm2::` and `utils::` should be documented (with doxygen-tags).
Everything else will probably just have a introduction above the class..

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+   ENDFOREACH(f)
+   SET(${var} "${listVar}" PARENT_SCOPE)
+ENDFUNCTION(PREPEND)
+
+PREPEND(LAYOUT_P "resources/ui" ${LAYOUT})
+add_custom_command(
+   COMMENT "Generate pot file"
+   TARGET ${PROJECT_NAME}
+   PRE_BUILD
+   COMMAND ${GETTEXT_XGETTEXT_EXECUTEABLE} -o "${CMAKE_CURRENT_SOURCE_DIR}/i18n/source.pot" --add-location=file --from-code=UTF-8 --sort-output --c++ --omit-header -k_ -s ${SOURCES}
```
Comment (Spagy) [inline src/CMakeLists.txt]: Why sort the output?

Comment (KoKuToru) [inline src/CMakeLists.txt]: If I remember correctly...
because the following merge command didn't work right without sorting.

File: src/i18n/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+foreach(_trans ${TRANSLATIONS})
+add_custom_command(
+   COMMENT "Generate Locale '${_trans}'"
+   OUTPUT "${CMAKE_CURRENT_SOURCE_DIR}/Locale/${_trans}.po"
+   POST_BUILD
+   COMMAND mkdir -p Locale/${_trans}/LC_MESSAGES
+   COMMAND ${GETTEXT_MSGATTRIB_EXECUTABLE} --translated "${CMAKE_CURRENT_SOURCE_DIR}/Locale/${_trans}.po" -o "${CMAKE_CURRENT_SOURCE_DIR}/Locale/${_trans}.po"
+   COMMAND ${GETTEXT_MSGMERGE_EXECUTABLE} --update --no-fuzzy-matching "${CMAKE_CURRENT_SOURCE_DIR}/Locale/${_trans}.po" "${CMAKE_CURRENT_SOURCE_DIR}/Locale/en.po"
```
Comment (Spagy) [inline src/i18n/CMakeLists.txt]: This looks as though it removes any translations that haven't also been translated into english. Why is that happening? Or have I misunderstood?

Comment (KoKuToru) [inline src/i18n/CMakeLists.txt]: `Line 30` removes _empty translations_ (untranslated) 
`Line 31` merges _missing translations_ from the `en.po`.

Without `Line 30` it would keep the _empty translations_

Which is normal no problem because the `msgid` would be the english-translation..
but isn't in our case, that's why we need to replace the empty translations with the right english translation.

Comment (KoKuToru) [inline src/i18n/CMakeLists.txt]: Oh you mean the `en.po`-case ?..
is no problem because
`Line 32` merges every _new/missing/empty translations_ into it from the `template.po`

Comment (Spagy) [inline src/i18n/CMakeLists.txt]: Could this be the intention?
msgcat --use-first %{_trans}.po en.po -o %{_trans}.po
So that any untranslated strings use the English if available?

Comment (KoKuToru) [inline src/i18n/CMakeLists.txt]: I did try `--use-first`, but did not work for me..

Comment (Spagy) [inline src/i18n/CMakeLists.txt]: " `Line 30` removes *empty translations* (untranslated)"
Looks good to me

"`Line 31` merges *missing translations* from the `en.po`"
Line 31 uses msgmerge
Here is the [description of msgmerge](https://www.gnu.org/software/gettext/manual/html_node/msgmerge-Invocation.html#msgmerge-Invocation). It states that any comments and translations form the second file are discarded.

Should msgcat be used in stead?
E.g. `msgcat --use-first fr.po en.po -o fr.po`
Of course this would have to be changed slightly for cmake

"`Line 32` merges every *new/missing/empty translations* into it from the `template.po`"
Okay, that seems right.

It adds empty translations to %{_trans}.po for each non existing translation. This is unnecessary if it is being generated to be used as input to a .mo file because empty translations are left out of the .mo file.

But this does makes sense if the .po file is being generated for translators as well, which I assume it is.

TL;DR
Should msgcat be used instead of msgmerge on L31?
Is L32 generating .po files for translators or are they made elsewhere?

Comment (KoKuToru) [inline src/i18n/CMakeLists.txt]: `Line 31` uses `--update` ... uhm whatever .. I have no idea

Yeah `Line 32` just makes sure the empty messages are in the po file if there isn't any en-translation.. shouldn't be needed if en is up to date always.. it got generated for the translators..

Anyways.. whatever .. do what you think is best because I have no idea how to work with translations.

You also can change the `*.ui` files to directly use english (instead of keys)
But I have no idea how to update translation in english without destroying other translation that reference to these english translations (as `msgid`)...

The way it's now, **it worked**-tm. I am not proud of this translation-mess

It just had 2 things todo
- make sure all new `msgid` are in (from \*.ui and \*.cpp/\*.h)
- make sure empty/missing translation have en-translation

`msgcat` and `msgmerge` behaved very different from what I imagined and this is the crazy **get it somehow working**-solution.

Comment (Spagy) [inline src/i18n/CMakeLists.txt]: Okay :)
I will play around with it. Could I just ask why you are okay with strings in .ui files but not source?

Comment (KoKuToru) [inline src/i18n/CMakeLists.txt]: Yes feel free to replace all texts from `.ui`-file.

While moving to `Gtk::Builder` i thought.. these keys are stupid...
Before the `Gtk::Builder` rewrite everything was in source...

Comment (Spagy) [inline src/i18n/CMakeLists.txt]: Sorry I might have been unclear. I don't understand why you don't like strings in the source. Why don't you like them?

Comment (KoKuToru) [inline src/i18n/CMakeLists.txt]: No I did have all strings in source because we created the window/dialogs at runtime,
that is why I used the key/value solution. (Too much text in source)

After some time I replaced all this written-code with `Gtk::Builder` using the `*.ui`-files.
Using the translation-keys in the `*.ui` in Glade is pretty ugly…

The keys in source are okay, they are only used for error-messages now.
All dialogs and widgets have their text now from `*.ui`

Comment (Spagy) [inline src/i18n/CMakeLists.txt]: Thank you for explaining. I would have them in source because it is easier to translate strings than keys but I think we have mostly solved that by making sure that new keys always have English translations.

Anyway, I am happy to have English strings in UI and keys in source; it shouldn't be too hard to make a build script for that.

**Comment by KoKuToru** [comment]
Just as a note, if somebody else worked on this branch.. the branch got rebased !

**Comment by KoKuToru** [comment]
Not 100% finish but not much missing, I will merge this now
